src/systemd/cockpit.service.in

```diff
@@ -3,14 +3,16 @@ Description=Cockpit Web Service
 Documentation=man:cockpit-ws(8)
 Requires=cockpit.socket
 Requires=cockpit-wsinstance-http.socket cockpit-wsinstance-https-factory.socket
-After=cockpit-wsinstance-http.socket cockpit-wsinstance-https-factory.socket
+# we need to start before the sockets so that the dynamic user exists
+Before=cockpit-wsinstance-http.socket cockpit-wsinstance-https-factory.socket
 
 [Service]
 RuntimeDirectory=cockpit/tls
 # systemd ≥ 241 sets this automatically
 Environment=RUNTIME_DIRECTORY=/run/cockpit/tls
 ExecStartPre=+@libexecdir@/cockpit-certificate-ensure --for-cockpit-tls
 ExecStart=@libexecdir@/cockpit-tls
+DynamicUser=true
```

Comment:
Should we set StateDirectory or ConfigurationDirectory here as well?
https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#RuntimeDirectory=

Reply:
To make sure that the user:group matches the dynamic user one.

Reply:
cockpit-tls doesn't actually write any state or config, it just reads /etc/cockpit (which is root:root). I did see that bit in the docs, but I didn't find it necessary. AFAICS, if it ever *does* start to write files, that would just fail on a permission error without setting a private StateDirectory, so we'd notice.

Reply:
I think this isn't 100% OK.  In my experimenting with this stuff before, one of the reasons that I stopped is that it was possible to get the system into a weird state where the `cockpit.service` process had exited (taking the dynamic user with it) but the `cockpit-wsinstance-*` stuff was not yet dead, resulting in the sockets in `/run/cockpit/wsinstance/` having ownership by a non-existing user.

Doing this in a race-free way would require the socket owner to be the one creating the dynamic group, and that's exactly the thing we requested from upstream systemd and ended up blocking on for so long...

There's discussion in https://github.com/systemd/systemd/issues/23067 that's still completely relevant, unless something has changed that I'm unaware of...

Reply:
One thing that has changed in the meantime is the introduction of the `OpenPath` option in systemd unit files.  We could definitely use this to open a restricted private directory that has world-accessible sockets in it and pass it to all the things that should have access to that directory.  Sort of like a private IPC area for all cockpit-related processes, if you will...

That would also let us move the session to socket activation... although I start to wonder if some variant of `systemd-run` could do this for us...

Reply:
@allisonkarlitskaya The cockpit-wsinstance*.socket units already have `BindsTo=cockpit.service`, so they will be stopped when you stop cockpit.service. (Just re-tested in a VM). Yes, they leave /run/cockpit/wsinstance/*.socket behind (which feels like a systemd bug), but they will be re-initialized once the unit starts again. I tried `chown dbus:qemu /run/cockpit/wsinstance/http*` after stopping cockpit.service, and the next request to http://localhost:9090 reinitializes them just fine.

So I don't see what's broken.

Reply:
So the last time I tested this, I ran into the issue with the stale sockets being left around and noticed them being owned by a numeric user ID and assumed that I had caused this problem with my hacks.  I tested an image yesterday with a clean install of cockpit from `main` and indeed also saw the stale socket issue (sockets there, but can't connect to them), so as you say, it isn't new.  I think the thing that was missing for me before was the knowledge that this is an already-existing problem.

It would help me a lot to swallow this if we had a solid systemd bug reference.

Reply:
I sent https://github.com/systemd/systemd/issues/32678 . If you generally warm up to the idea, I'll add it as a comment and add the integration test case for restoring permissions.

Reply:
Sounds like we have the proper solution to this, thanks to Lennart...

Reply:
Applied the RemoveOnStop= in a separate commit with a test case.